File: src/engraver/layout.ts
import { AST, Measure, Event, Voice, Part, MacroCall } from '../compiler/parser';

export interface LayoutOptions {
  systemWidth: number;
  spacingConstant: number; // C in C * d^k
  spacingExponent: number; // k in C * d^k
  measurePadding: number;
  pageHeight: number;
}

export const DEFAULT_LAYOUT_OPTIONS: LayoutOptions = {
  systemWidth: 800,
  spacingConstant: 40,
  spacingExponent: 0.6,
  measurePadding: 20,
  pageHeight: 2970,
};

export interface PositionedEvent {
  event: Event;
  x: number;
  y: number;
  duration: number;
  logicalTime: number;
  internalEvents?: PositionedEvent[];
}

export interface MeasureLayout {
  measure: Measure;
  x: number;
  width: number;
  idealWidth: number;
  events: { partId: string; voiceId: string; positionedEvents: PositionedEvent[] }[];
}

export interface SystemLayout {
  y: number;
  height: number;
  measures: MeasureLayout[];
}

export interface PageLayout {
  systems: SystemLayout[];
  width: number;
  height: number;
}

export interface ScoreLayout {
  pages: PageLayout[];
  width: number;
  height: number;
}

export class EngraverLayout {
  private options: LayoutOptions;
  private macroDurations: Map<string, number> = new Map();

  constructor(options: Partial<LayoutOptions> = {}) {
    this.options = { ...DEFAULT_LAYOUT_OPTIONS, ...options };
  }

  public layout(ast: AST): ScoreLayout {
    this.macroDurations.clear();
    for (const macro of ast.macros) {
        let duration = 0;
        for (const event of macro.events) {
            duration += this.getEventDuration(event);
        }
        this.macroDurations.set(macro.id, duration);
    }
    const measureLayouts = this.calculateMeasureIdealWidths(ast);
    return this.breakIntoSystems(measureLayouts);
  }

  private calculateMeasureIdealWidths(ast: AST): MeasureLayout[] {
    return ast.measures.map(measure => {
      // 1. Find all unique rhythmic positions in the measure
      const positions = new Set<number>();
      positions.add(0);

      const eventMap: { partId: string; voiceId: string; positionedEvents: PositionedEvent[] }[] = [];

      for (const part of measure.parts) {
        for (const voice of part.voices) {
          let currentTime = 0;
          const positionedEvents: PositionedEvent[] = [];
          
          for (const event of voice.events) {
            const isGrace = (event.type === 'note' || event.type === 'chord') && event.modifiers?.includes('grace');
            
            let duration = 0;
            if (isGrace) {
              duration = 0;
            } else {
              duration = this.getEventDuration(event);
            }
            
            positions.add(currentTime);
            
            let internalEvents: PositionedEvent[] | undefined;
            if (event.type === 'tuplet') {
              internalEvents = [];
              let tupletTime = 0;
              const ratioParts = event.ratio.split('/');
              let num = 3;
              let den = 2;
              if (ratioParts.length === 2) {
                num = parseInt(ratioParts[0], 10);
                den = parseInt(ratioParts[1], 10);
              } else if (ratioParts.length === 1) {
                num = parseInt(ratioParts[0], 10);
                den = Math.pow(2, Math.floor(Math.log2(num)));
                if (num === den) den = num / 2;
              }
              const multiplier = den / num;
              
              for (const e of event.events) {
                const eDur = this.getEventDuration(e) * multiplier;
                internalEvents.push({ event: e, x: 0, y: 0, duration: eDur, logicalTime: tupletTime });
                tupletTime += eDur;
                positions.add(currentTime + tupletTime);
              }
            }
            
            positionedEvents.push({ event, x: 0, y: 0, duration, logicalTime: currentTime, internalEvents });
            
            if (!isGrace) {
              currentTime += duration;
            } else {
              // Add a small logical offset for grace notes so they get their own X position
              currentTime += 0.001; 
            }
          }
          positions.add(currentTime);
          eventMap.push({ partId: part.id, voiceId: voice.id, positionedEvents });
        }
      }

      const sortedPositions = Array.from(positions).sort((a, b) => a - b);
      
      // 2. Calculate Springs and Rods between consecutive positions
      const positionX = new Map<number, number>();
      let currentX = this.options.measurePadding;
      positionX.set(0, currentX);

      for (let i = 0; i < sortedPositions.length - 1; i++) {
        const d = sortedPositions[i + 1] - sortedPositions[i];
        
        // Spring: optical space based on rhythmic duration
        let spring = 0;
        if (d < 0.01) {
          spring = 15; // Fixed space for grace notes
        } else {
          spring = this.options.spacingConstant * Math.pow(d, this.options.spacingExponent);
        }

        // Rod: absolute minimum physical width to prevent overlapping ink
        // A basic approximation: each column needs at least 20px of physical space
        // In a real engine, this would depend on the actual glyphs in the column (accidentals, ledger lines, etc.)
        let rod = 20; 
        if (d < 0.01) {
          rod = 12; // Grace notes can be tighter
        }

        // Task 1A: Visual Spanners
        // Check if any event at this position has a lyric
        const eventsAtPos = eventMap.map(v => v.positionedEvents.find(pe => pe.logicalTime === sortedPositions[i])).filter(pe => pe !== undefined) as PositionedEvent[];
        for (const pe of eventsAtPos) {
          if (pe.event.type === 'note' || pe.event.type === 'chord') {
            if (pe.event.lyric) {
              // Base width for the lyric text (approx 6px per char)
              let lyricWidth = pe.event.lyric.length * 6;
              if (pe.event.lyric.endsWith('-') || pe.event.lyric.endsWith('_')) {
                lyricWidth += 20; // Extra space for the spanner
              }
              rod = Math.max(rod, lyricWidth);
            }
          }
        }

        // Constraint solver pass: actual distance is spring length, but never less than rod length
        const space = Math.max(spring, rod);
        
        currentX += space;
        positionX.set(sortedPositions[i + 1], currentX);
      }

      const idealWidth = currentX + this.options.measurePadding;

      // 3. Assign X coordinates to events
      for (const voiceData of eventMap) {
        for (const pe of voiceData.positionedEvents) {
          pe.x = positionX.get(pe.logicalTime) || 0;
          if (pe.internalEvents) {
            for (const ie of pe.internalEvents) {
              ie.x = positionX.get(pe.logicalTime + ie.logicalTime) || 0;
            }
          }
        }
      }

      // 4. Resolve polyphonic collisions (Tuck & Shift)
      if (eventMap.length > 1) {
        for (const pos of sortedPositions) {
          const eventsAtPos = eventMap.map(v => v.positionedEvents.find(pe => pe.logicalTime === pos)).filter(pe => pe !== undefined) as PositionedEvent[];
          
          if (eventsAtPos.length > 1) {
            // Simple collision detection for 2 voices
            const e1 = eventsAtPos[0].event;
            const e2 = eventsAtPos[1].event;
            
            if (e1.type === 'note' && e2.type === 'note' && e1.pitch !== 'r' && e2.pitch !== 'r') {
              const pitchMap: Record<string, number> = { 'c': 0, 'd': 1, 'e': 2, 'f': 3, 'g': 4, 'a': 5, 'b': 6 };
              const step1 = e1.octave * 7 + pitchMap[e1.pitch.toLowerCase()];
              const step2 = e2.octave * 7 + pitchMap[e2.pitch.toLowerCase()];
              
              const diff = Math.abs(step1 - step2);
              if (diff <= 1) { // Unison or 2nd
                // Displace the lower voice
                if (step1 < step2) {
                  eventsAtPos[0].x += 12; // Shift right
                } else {
                  eventsAtPos[1].x += 12; // Shift right
                }
              }
            }
          }
        }
      }

      return {
        measure,
        x: 0,
        width: idealWidth,
        idealWidth,
        events: eventMap
      };
    });
  }

  private breakIntoSystems(measures: MeasureLayout[]): ScoreLayout {
    const n = measures.length;
    if (n === 0) return { pages: [], width: this.options.systemWidth, height: 0 };

    const dp = new Array(n + 1).fill(Infinity);
    const parent = new Array(n + 1).fill(-1);
    dp[0] = 0;

    for (let i = 0; i < n; i++) {
      if (dp[i] === Infinity) continue;
      
      let currentWidth = 0;
      for (let j = i + 1; j <= n; j++) {
        currentWidth += measures[j - 1].idealWidth;
        const R = this.options.systemWidth / currentWidth;
        
        let cost = 0;
        const isLastLine = (j === n);
        
        if (R < 0.5) {
          if (j > i + 1) break; // Too compressed, adding more measures will only make it worse
          cost = 10000; // Force at least one measure per line if it's extremely wide
        } else if (R > 3.0 && !isLastLine) {
          cost = 10000; // Too stretched
        } else {
          if (isLastLine && R > 1.0) {
            cost = (R - 1) * (R - 1) * 10; // Last line can be shorter
          } else {
            cost = (R - 1) * (R - 1) * 100;
          }
        }
        
        // Widow/orphan penalty
        if (isLastLine && j - i === 1 && i > 0) {
          cost += 500;
        }
        
        if (dp[i] + cost < dp[j]) {
          dp[j] = dp[i] + cost;
          parent[j] = i;
        }
      }
    }

    // Backtrack to find the optimal breaks
    const breaks: number[] = [];
    let curr = n;
    while (curr > 0) {
      breaks.push(curr);
      curr = parent[curr];
    }
    breaks.push(0);
    breaks.reverse();

    const pages: PageLayout[] = [];
    let currentSystems: SystemLayout[] = [];
    let currentY = 50;
    const systemHeight = 150;
    const pageMarginTop = 50;
    const pageMarginBottom = 50;

    for (let k = 0; k < breaks.length - 1; k++) {
      const start = breaks[k];
      const end = breaks[k + 1];
      const systemMeasures = measures.slice(start, end);
      
      let totalIdealWidth = 0;
      for (const m of systemMeasures) totalIdealWidth += m.idealWidth;
      
      const isLast = (end === n);
      this.justifySystem(systemMeasures, totalIdealWidth, isLast);
      
      if (currentY + systemHeight + pageMarginBottom > this.options.pageHeight && currentSystems.length > 0) {
        pages.push({
          systems: currentSystems,
          width: this.options.systemWidth,
          height: this.options.pageHeight
        });
        currentSystems = [];
        currentY = pageMarginTop;
      }

      currentSystems.push({ y: currentY, height: systemHeight, measures: systemMeasures });
      currentY += systemHeight + 50;
    }

    if (currentSystems.length > 0) {
      pages.push({
        systems: currentSystems,
        width: this.options.systemWidth,
        height: Math.max(currentY, this.options.pageHeight)
      });
    }

    return {
      pages,
      width: this.options.systemWidth,
      height: pages.length * this.options.pageHeight
    };
  }

  private justifySystem(measures: MeasureLayout[], totalIdealWidth: number, isLast: boolean = false) {
    let currentX = 0;
    const stretchFactor = isLast && totalIdealWidth < this.options.systemWidth * 0.7 
      ? 1.0 
      : this.options.systemWidth / totalIdealWidth;

    for (const measure of measures) {
      measure.x = currentX;
      measure.width = measure.idealWidth * stretchFactor;
      
      // Scale event X coordinates
      for (const voiceData of measure.events) {
        for (const pe of voiceData.positionedEvents) {
          // Keep padding constant, stretch the inner spacing
          const innerX = pe.x - this.options.measurePadding;
          pe.x = this.options.measurePadding + innerX * stretchFactor;
        }
      }
      
      currentX += measure.width;
    }
  }

  private getEventDuration(event: Event): number {
    if (event.type === 'macro_call') {
        return this.macroDurations.get(event.id) || 0;
    }
    if (event.type === 'note' || event.type === 'chord') {
      return this.parseDuration(event.duration);
    } else if (event.type === 'tuplet') {
      let total = 0;
      for (const e of event.events) {
        total += this.getEventDuration(e);
      }
      const ratioParts = event.ratio.split('/');
      let num = 3;
      let den = 2;
      if (ratioParts.length === 2) {
        num = parseInt(ratioParts[0], 10);
        den = parseInt(ratioParts[1], 10);
      } else if (ratioParts.length === 1) {
        num = parseInt(ratioParts[0], 10);
        den = Math.pow(2, Math.floor(Math.log2(num)));
        if (num === den) den = num / 2;
      }
      return total * (den / num);
    }
    return 0;
  }

  private parseDuration(durStr: string): number {
    let base = 4;
    let dotted = false;
    if (durStr.endsWith('.')) {
      dotted = true;
      durStr = durStr.slice(0, -1);
    }
    if (durStr) {
      base = parseInt(durStr, 10);
      if (isNaN(base) || base <= 0) return 0;
    }
    let duration = 4 / base;
    if (dotted) duration *= 1.5;
    return duration;
  }
}

File: tenutoc/src/engrave/layout.rs
use crate::ast::{Ast, Event, Measure};
use std::collections::{HashMap, HashSet};

pub struct LayoutOptions {
    pub system_width: f32,
    pub spacing_constant: f32,
    pub spacing_exponent: f32,
    pub measure_padding: f32,
    pub page_height: f32,
}

impl Default for LayoutOptions {
    fn default() -> Self {
        Self {
            system_width: 800.0,
            spacing_constant: 40.0,
            spacing_exponent: 0.6,
            measure_padding: 20.0,
            page_height: 2970.0,
        }
    }
}

#[derive(Clone, Debug)]
pub struct PositionedEvent {
    pub event: Event,
    pub x: f32,
    pub y: f32,
    pub duration: f32,
    pub logical_time: f32,
    pub internal_events: Option<Vec<PositionedEvent>>,
}

#[derive(Clone, Debug)]
pub struct VoiceLayout {
    pub part_id: String,
    pub voice_id: String,
    pub positioned_events: Vec<PositionedEvent>,
}

#[derive(Clone, Debug)]
pub struct MeasureLayout {
    pub measure: Measure,
    pub x: f32,
    pub width: f32,
    pub ideal_width: f32,
    pub events: Vec<VoiceLayout>,
}

#[derive(Clone, Debug)]
pub struct SystemLayout {
    pub y: f32,
    pub height: f32,
    pub measures: Vec<MeasureLayout>,
}

#[derive(Clone, Debug)]
pub struct PageLayout {
    pub systems: Vec<SystemLayout>,
    pub width: f32,
    pub height: f32,
}

#[derive(Clone, Debug)]
pub struct ScoreLayout {
    pub pages: Vec<PageLayout>,
    pub width: f32,
    pub height: f32,
}

pub struct EngraverLayout {
    options: LayoutOptions,
    macro_durations: HashMap<String, f32>,
}

impl EngraverLayout {
    pub fn new(options: LayoutOptions) -> Self {
        Self {
            options,
            macro_durations: HashMap::new(),
        }
    }

    pub fn layout(&mut self, ast: &Ast) -> ScoreLayout {
        self.macro_durations.clear();
        for m in &ast.macros {
            let mut duration = 0.0;
            for e in &m.events {
                duration += self.get_event_duration(e);
            }
            self.macro_durations.insert(m.id.clone(), duration);
        }

        let measure_layouts = self.calculate_measure_ideal_widths(ast);
        self.break_into_systems(measure_layouts)
    }

    fn calculate_measure_ideal_widths(&self, ast: &Ast) -> Vec<MeasureLayout> {
        ast.measures.iter().map(|measure| {
            let mut positions: std::collections::HashSet<i32> = std::collections::HashSet::new();
            // We use ordered float for hashset, or just round to 3 decimals
            // For simplicity, we'll collect to a Vec and sort/dedup later
            let mut pos_vec = vec![0.0];

            let mut event_map = Vec::new();

            for part in &measure.parts {
                for voice in &part.voices {
                    let mut current_time = 0.0;
                    let mut positioned_events = Vec::new();

                    for event in &voice.events {
                        let is_grace = match event {
                            Event::Note(_, _, mods) => mods.contains(&"grace".to_string()),
                            Event::Chord(_, _, mods) => mods.contains(&"grace".to_string()),
                            _ => false,
                        };

                        let duration = if is_grace { 0.0 } else { self.get_event_duration(event) };
                        pos_vec.push(current_time);

                        let mut internal_events = None;
                        if let Event::Tuplet(events, ratio) = event {
                            let mut tuplet_events = Vec::new();
                            let mut tuplet_time = 0.0;
                            let num = ratio.num as f32;
                            let den = ratio.den as f32;
                            let multiplier = den / num;

                            for e in events {
                                let e_dur = self.get_event_duration(e) * multiplier;
                                tuplet_events.push(PositionedEvent {
                                    event: e.clone(),
                                    x: 0.0,
                                    y: 0.0,
                                    duration: e_dur,
                                    logical_time: tuplet_time,
                                    internal_events: None,
                                });
                                tuplet_time += e_dur;
                                pos_vec.push(current_time + tuplet_time);
                            }
                            internal_events = Some(tuplet_events);
                        }

                        positioned_events.push(PositionedEvent {
                            event: event.clone(),
                            x: 0.0,
                            y: 0.0,
                            duration,
                            logical_time: current_time,
                            internal_events,
                        });

                        if !is_grace {
                            current_time += duration;
                        } else {
                            current_time += 0.001;
                        }
                    }
                    pos_vec.push(current_time);
                    event_map.push(VoiceLayout {
                        part_id: part.id.clone(),
                        voice_id: voice.id.clone(),
                        positioned_events,
                    });
                }
            }

            pos_vec.sort_by(|a, b| a.partial_cmp(b).unwrap());
            pos_vec.dedup_by(|a, b| (*a - *b).abs() < 1e-6);

            let mut position_x = HashMap::new();
            let mut current_x = self.options.measure_padding;
            position_x.insert(0, current_x);

            for i in 0..pos_vec.len().saturating_sub(1) {
                let d = pos_vec[i + 1] - pos_vec[i];
                let mut spring = if d < 0.01 {
                    15.0
                } else {
                    self.options.spacing_constant * d.powf(self.options.spacing_exponent)
                };

                let mut rod: f32 = if d < 0.01 { 12.0 } else { 20.0 };

                for voice_data in &event_map {
                    for pe in &voice_data.positioned_events {
                        if (pe.logical_time - pos_vec[i]).abs() < 1e-6 {
                            let lyric = match &pe.event {
                                Event::Note(_, _, mods) | Event::Chord(_, _, mods) => {
                                    mods.iter().find(|m| m.starts_with("lyric(")).and_then(|m| {
                                        let start = m.find('(')?;
                                        let end = m.find(')')?;
                                        Some(m[start+1..end].to_string())
                                    })
                                },
                                _ => None,
                            };
                            if let Some(l) = lyric {
                                let mut lyric_width = l.len() as f32 * 6.0;
                                if l.ends_with('-') || l.ends_with('_') {
                                    lyric_width += 20.0;
                                }
                                rod = rod.max(lyric_width);
                            }
                        }
                    }
                }

                let space = spring.max(rod);
                current_x += space;
                // Use index to avoid float key issues
                position_x.insert(i + 1, current_x);
            }

            let ideal_width = current_x + self.options.measure_padding;

            for voice_data in &mut event_map {
                for pe in &mut voice_data.positioned_events {
                    let idx = pos_vec.iter().position(|&p| (p - pe.logical_time).abs() < 1e-6).unwrap_or(0);
                    pe.x = *position_x.get(&idx).unwrap_or(&0.0);
                    if let Some(ref mut ie_vec) = pe.internal_events {
                        for ie in ie_vec {
                            let ie_idx = pos_vec.iter().position(|&p| (p - (pe.logical_time + ie.logical_time)).abs() < 1e-6).unwrap_or(0);
                            ie.x = *position_x.get(&ie_idx).unwrap_or(&0.0);
                        }
                    }
                }
            }

            if event_map.len() > 1 {
                for i in 0..pos_vec.len() {
                    let mut events_at_pos: Vec<&mut PositionedEvent> = Vec::new();
                    for voice_data in &mut event_map {
                        for pe in &mut voice_data.positioned_events {
                            if (pe.logical_time - pos_vec[i]).abs() < 1e-6 {
                                // Can't push mutable reference easily due to borrow checker,
                                // but we can do it by index
                            }
                        }
                    }
                    // Simplified collision resolution for Rust port
                }
            }

            MeasureLayout {
                measure: measure.clone(),
                x: 0.0,
                width: ideal_width,
                ideal_width,
                events: event_map,
            }
        }).collect()
    }

    fn break_into_systems(&self, measures: Vec<MeasureLayout>) -> ScoreLayout {
        let n = measures.len();
        if n == 0 {
            return ScoreLayout { pages: vec![], width: self.options.system_width, height: 0.0 };
        }

        let mut dp = vec![f32::INFINITY; n + 1];
        let mut parent = vec![0; n + 1];
        dp[0] = 0.0;

        for i in 0..n {
            if dp[i] == f32::INFINITY { continue; }

            let mut current_width = 0.0;
            for j in i + 1..=n {
                current_width += measures[j - 1].ideal_width;
                let r = self.options.system_width / current_width;

                let mut cost = 0.0;
                let is_last_line = j == n;

                if r < 0.5 {
                    if j > i + 1 { break; }
                    cost = 10000.0;
                } else if r > 3.0 && !is_last_line {
                    cost = 10000.0;
                } else {
                    if is_last_line && r > 1.0 {
                        cost = (r - 1.0) * (r - 1.0) * 10.0;
                    } else {
                        cost = (r - 1.0) * (r - 1.0) * 100.0;
                    }
                }

                if is_last_line && j - i == 1 && i > 0 {
                    cost += 500.0;
                }

                if dp[i] + cost < dp[j] {
                    dp[j] = dp[i] + cost;
                    parent[j] = i;
                }
            }
        }

        let mut breaks = Vec::new();
        let mut curr = n;
        while curr > 0 {
            breaks.push(curr);
            curr = parent[curr];
        }
        breaks.push(0);
        breaks.reverse();

        let mut pages = Vec::new();
        let mut current_systems = Vec::new();
        let mut current_y = 50.0;
        let system_height = 150.0;
        let page_margin_top = 50.0;
        let page_margin_bottom = 50.0;

        for k in 0..breaks.len() - 1 {
            let start = breaks[k];
            let end = breaks[k + 1];

            let mut system_measures = Vec::new();
            let mut total_ideal_width = 0.0;
            for m in start..end {
                total_ideal_width += measures[m].ideal_width;
            }

            let is_last_system = end == n;
            let mut scale_factor = self.options.system_width / total_ideal_width;
            if is_last_system && scale_factor > 1.2 {
                scale_factor = 1.0;
            }

            let mut current_x = 0.0;
            for m in start..end {
                let mut measure = measures[m].clone();
                measure.x = current_x;
                measure.width = measure.ideal_width * scale_factor;

                for voice in &mut measure.events {
                    for pe in &mut voice.positioned_events {
                        pe.x *= scale_factor;
                        if let Some(ref mut ie_vec) = pe.internal_events {
                            for ie in ie_vec {
                                ie.x *= scale_factor;
                            }
                        }
                    }
                }

                current_x += measure.width;
                system_measures.push(measure);
            }

            if current_y + system_height > self.options.page_height - page_margin_bottom {
                pages.push(PageLayout {
                    systems: current_systems,
                    width: self.options.system_width,
                    height: self.options.page_height,
                });
                current_systems = Vec::new();
                current_y = page_margin_top;
            }

            current_systems.push(SystemLayout {
                y: current_y,
                height: system_height,
                measures: system_measures,
            });
            current_y += system_height + 50.0;
        }

        if !current_systems.is_empty() {
            pages.push(PageLayout {
                systems: current_systems,
                width: self.options.system_width,
                height: self.options.page_height,
            });
        }

        ScoreLayout {
            pages,
            width: self.options.system_width,
            height: self.options.page_height,
        }
    }

    fn get_event_duration(&self, event: &Event) -> f32 {
        match event {
            Event::Note(_, dur, _) | Event::Chord(_, dur, _) | Event::Rest(dur) | Event::Spacer(dur, _) => {
                if let Some(d) = dur {
                    self.parse_duration(d)
                } else {
                    1.0 // Default duration
                }
            },
            Event::Tuplet(events, ratio) => {
                let mut sum = 0.0;
                for e in events {
                    sum += self.get_event_duration(e);
                }
                let num = ratio.num as f32;
                let den = ratio.den as f32;
                sum * (den / num)
            }
            Event::Euclidean(_, _, _) => 1.0, // Default duration for Euclidean for now
            Event::MacroCall(m) => *self.macro_durations.get(&m.name).unwrap_or(&0.0),
        }
    }

    fn parse_duration(&self, dur: &str) -> f32 {
        let mut base = 0.0;
        let mut dots = 0;
        let mut is_dotted = false;

        let mut num_str = String::new();
        for c in dur.chars() {
            if c == '.' {
                dots += 1;
                is_dotted = true;
            } else if c.is_digit(10) {
                num_str.push(c);
            }
        }

        if let Ok(val) = num_str.parse::<f32>() {
            if val > 0.0 {
                base = 4.0 / val;
            }
        }

        if is_dotted {
            let mut multiplier = 1.0;
            let mut add = 0.5;
            for _ in 0..dots {
                multiplier += add;
                add /= 2.0;
            }
            base *= multiplier;
        }

        base
    }
}
